### PR TITLE
A github template for Sceptre template handlers

### PR DESCRIPTION
**Pull request by zaro0508**
Initial creation of a Github template that will be used to create
template handler repos.  This GH template will setup the following

* setup pre-commit to run linters on every commit
* files to help geerate a template handler
* setup the CI workflow to run tests
* setup the CI workflow to build and deploy the packages to pypi
* provide a template handler skeleton for code and docs

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    docker:
+      - image: cloudreach/sceptre-circleci:0.8.0
+
+    steps:
+      - checkout
+
+      - restore_cache:
+          key:
+            v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile"
+            }}
```
Comment (jfalkenstein) [inline .circleci/config.yml]: Why are these curly braces on the next line? Seems like they should just be moved back up

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+          command: |
+            python3 -m venv venv
+            . venv/bin/activate
+            pip3 install --upgrade pip
+            pip3 install -r requirements.txt
+
+      - save_cache:
+          key:
+            v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile"
+            }}
```
Comment (jfalkenstein) [inline .circleci/config.yml]: Same comment here.

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            pip3 install -e .
+
+      - persist_to_workspace:
+          root: /home/circleci
+          paths:
+            - project
+
+  lint-and-unit-test:
+    docker:
+      - image: cloudreach/sceptre-circleci:0.7.0
```
Comment (jfalkenstein) [inline .circleci/config.yml]: Why is this a different version than the one used at the top?

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    steps:
+      - attach_workspace:
+          at: /home/circleci
+
+      - checkout
+
+      - restore_cache:
+          key:
+            v1-dependency-cache-{{ checksum "setup.py" }}-{{ checksum "Makefile"
+            }}
```
Comment (jfalkenstein) [inline .circleci/config.yml]: This seems copied throughout this file, but I think there's an extra linebreak each time. I'll stop commenting on it, but I think it should be fixed everywhere.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+          destination: coverage-reports
+
+      - persist_to_workspace:
+          root: /home/circleci
+          paths:
+            - project
+
+  deploy-test:
+    docker:
+      - image: cloudreach/sceptre-circleci:0.7.0
```
Comment (jfalkenstein) [inline .circleci/config.yml]: This too is a different version than the one at top.

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+          paths:
+            - 'venv'
+
+      - run:
+          name: Initialise .pypirc
+          command: |
+            echo -e "[distutils]\n  index-servers =\n    test-pypi\n\n" >> ~/.pypirc
+            echo -e "[test-pypi]" >> ~/.pypirc
+            echo -e "username: $PYPI_TEST_USER" >> ~/.pypirc
+            echo -e "password: $PYPI_TEST_PASSWORD" >> ~/.pypirc
```
Comment (jfalkenstein) [inline .circleci/config.yml:135]: Why are these TEST envs?

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+# if multiple use single string with commas.
+TEMPLATE_HANDLER_AUTHOR_EMAIL = 'sceptre@sceptre.org'
+TEMPLATE_HANDLER_URL = 'https://github.com/sceptre/{}'.format(TEMPLATE_HANDLER_NAME)
+
+with open("README.md") as readme_file:
+    README = readme_file.read()
+
+install_requirements = [
+    "packaging==16.8",
+    "requests>=2.25,<3"
```
Comment (jfalkenstein) [inline setup.py]: These seem odd to me. Why are these required on every template handler project?

File: template_handler/handler.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    Parameters
+    ----------
+    name: str
+        The name of the template. Corresponds to the name of the Stack this template belongs to.
+    handler_config: dict
+        Configuration of the template handler. All properties except for `type` are available.
+    sceptre_user_data: dict
+        Sceptre user data defined in the Stack config
+    connection_manager: sceptre.connection_manager.ConnectionManager
+        Connection manager that can be used to call AWS APIs
```
Comment (jfalkenstein) [inline template_handler/handler.py]: You forgot stack_group_config

File: template_handler/handler.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(CustomTemplateHandler, self).__init__(*args, **kwargs)
```
Comment (jfalkenstein) [inline template_handler/handler.py]: This really isn't necessary. You could remove this bit.

**Review by jfalkenstein — CHANGES_REQUESTED** [review]
This looks really good. I suspect there are some things that made their way into this project by virtue of copy/paste and they don't actually belong in it. I've commented on those places where I've seen them.

File: requirements.txt [hunk 10]
```diff
@@ -0,0 +1,13 @@
+bumpversion==0.5.3
+coverage==4.4.2
+pre-commit>=2.12.0,<2.13
+mock==2.0.0
```
Comment (jfalkenstein) [inline requirements.txt]: This is in the py3 stdlib now. No need to install it on its own.

File: requirements.txt [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+packaging==16.8
+pytest-runner>=3.0.0,<3.1.0
+pytest>=3.2.0,<3.3.0
+readme-renderer>=24.0
+setuptools>=40.6.2
+git+git://github.com/sceptre/sceptre.git@master#egg=sceptre
```
Comment (jfalkenstein) [inline requirements.txt]: Why install Sceptre's master? Why not just say `sceptre>=2.7`?

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    license='Apache2',
+    url=TEMPLATE_HANDLER_URL,
+    packages=find_packages(
+        exclude=["*.tests", "*.tests.*", "tests.*", "tests"]),
+    py_modules=[TEMPLATE_HANDLER_MODULE_NAME],
+    entry_points={
+        'sceptre.template_handlers': [
+            "{}={}:{}".format(TEMPLATE_HANDLER_COMMAND_NAME,
+                              TEMPLATE_HANDLER_MODULE_NAME, TEMPLATE_HANDLER_CLASS)
+        ]
```
Comment (jfalkenstein) [inline setup.py]: f-strings would make this better

File: setup.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+__version__ = "0.1.0"
+
+# More information on setting values:
+# https://github.com/Sceptre/project/wiki/sceptre-template-handler-template
+
+# lowercase, use `-` as separator.
+TEMPLATE_HANDLER_NAME = 'custom-template-handler'
+# the template_handler call in sceptre e.g. !command_name.
+TEMPLATE_HANDLER_COMMAND_NAME = 'custom_template_handler'
```
Comment (jfalkenstein) [inline setup.py]: This isn't how template handlers are used. Perhaps
```suggestion
# the template_handler call in sceptre e.g. !command_name.
TEMPLATE_TYPE = 'custom_template_handler'
```

File: setup.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        ]
+    },
+    include_package_data=True,
+    zip_safe=False,
+    keywords="sceptre, sceptre-template-handler",
+    classifiers=[
+        "Intended Audience :: Developers",
+        "Natural Language :: English",
+        "Environment :: Console",
+        "Programming Language :: Python :: 3.8"
```
Comment (jfalkenstein) [inline setup.py:64]: Only 3.8?

File: Makefile [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+coverage: coverage-all
+		coverage report --show-missing
+
+test:
+	    python -m pytest --junitxml=test-reports/junit.xml
+lint:
+	    pre-commit run --all-files --show-diff-on-failure
+
+acceptance-test:
+	    behave acceptance-tests/
```
Comment (jfalkenstein) [inline Makefile]: Neither behave nor the acceptance-tests directory exist in this repo.

File: .circleci/config.yml (hunk 1 above)
Comment (zaro0508) [reply .circleci/config.yml]: fixed

File: .circleci/config.yml (hunk 2 above)
Comment (zaro0508) [reply .circleci/config.yml]: fixed

File: .circleci/config.yml (hunk 3 above)
Comment (zaro0508) [reply .circleci/config.yml]: fixed

File: .circleci/config.yml (hunk 4 above)
Comment (zaro0508) [reply .circleci/config.yml]: fixed

File: .circleci/config.yml (hunk 5 above)
Comment (zaro0508) [reply .circleci/config.yml]: fixed

File: setup.py (hunk 7 above)
Comment (zaro0508) [reply setup.py]: removed.

File: template_handler/handler.py (hunk 9 above)
Comment (zaro0508) [reply template_handler/handler.py]: it's in the docs so i'll just leave as-is unless you feel strongly about removing it.

File: requirements.txt (hunk 10 above)
Comment (zaro0508) [reply requirements.txt]: removed

File: setup.py (hunk 12 above)
Comment (zaro0508) [reply setup.py]: changed to fstring

File: setup.py (hunk 13 above)
Comment (zaro0508) [reply setup.py]: done.

File: setup.py (hunk 14 above)
Comment (zaro0508) [reply setup.py:64]: i'm going to assume whoever uses this template to knows that they might want to update this section.

File: Makefile (hunk 15 above)
Comment (zaro0508) [reply Makefile]: removed

File: requirements.txt (hunk 11 above)
Comment (zaro0508) [reply requirements.txt]: I'm thinking that it might be better to not pin here just to make it less of a maintenance burden.

File: template_handler/handler.py (hunk 8 above)
Comment (zaro0508) [reply template_handler/handler.py]: fixed

File: requirements.txt (hunk 11 above)
Comment (jfalkenstein) [reply requirements.txt]: That's not a pin, it's just a version requirement. It wouldn't be appropriate for hooks or resolvers, but template handlers simply won't work prior to version 2.7.0. So anything greater than 2.7.0 is appropriate in the case of template handlers.

**Review by jfalkenstein — APPROVED** [review]

File: requirements.txt (hunk 11 above)
Comment (zaro0508) [reply requirements.txt]: sorry, i misread your previous comment.  I have change to require sceptre from pypi.
